lib/matplotlib/dates.py

```diff
@@ -424,13 +439,64 @@ def date2num(d):
             # make datetime naive:
             d = [dt.astimezone(UTC).replace(tzinfo=None) for dt in d]
             d = np.asarray(d)
-        d = d.astype('datetime64[us]')
+
+        # try conversion to datetime or timedelta
+        for _dtype in ('datetime64[us]', 'timedelta64[us]'):
+            try:
+                d = d.astype(_dtype)
```

Comment:
All of the following work:
```
np.array((datetime.datetime(2021, 1, 10),)).astype('datetime64[us]')
np.array((datetime.timedelta(days=1),)).astype('timedelta64[us]')

np.array((pandas.Timestamp('2017-01-01T12:13'),)).astype('datetime64[us]')
np.array((pandas.Timedelta(days=1),)).astype('timedelta64[us]')
```

And we could do `isinstance(d[0], datetime.datetime)`. But we can't do this for the two pandas datatypes as we can't import pandas. https://github.com/matplotlib/matplotlib/issues/4000#issuecomment-69948678

Therefore the try... except block. We can't check if the data is Timestamp or Timedelta, so we try it.
This will work for all cases except if the data contains `pandas.NaT` or is completely unsupported. Only then the else block will be used. There's no way to check if an object is `pandas.NaT` without importing pandas to do that apparently https://github.com/pandas-dev/pandas/issues/9253#issuecomment-69966529

Your example with `np.timedelta64` should never reach this part of the code.

Reply:
Sorry I just meant can you provide *me* with an example that fails?  I'm not familiar enough w pandas to know exactly what you are trying to do.  

Overall I spent quite a bit of effort simplifying this code.  If at all possible it would be nice to keep it simple.

Reply:
Yes, sorry. I'm not sure I understand what you mean by an example that "fails". Do you mean data that cannot be converted with .astype? Then an example would be:

```
import pandas as pd

# doesn't work
td = (pd.Timedelta(days=1), pd.Timedelta(days=2), pd.NaT)
date2num(td)

# doesn't work either
date = (pandas.Timestamp(year=2020, month=1, day=10),
        pandas.Timestamp(year=2020, month=1, day=11),
        pd.NaT)
date2num(date)
```

PS: sorry for the delay, I responded (or thought I did) and managed not to respond at the same time...

Reply:
Thanks!  

> PS: sorry for the delay, I responded (or thought I did) and managed not to respond at the same time...

Happens to me all the time - so many excellent email responses in the Drafts folder for months.

Reply:
I also had some better ideas for doing this.
```suggestion
                if isinstance(d.take(0), datetime.date):  # test first element
                    d = d.astype('datetime64[us]')
                else:
                    d = d.astype('timedelta64[us]')
```

Pandas Timestamp and Timedelta are subclasses of datetime.datetime and datetime.timedelta respectively so they can actually be handled by this code. Only pandas.NaT can't.

But, by using the same method as used already at the beginning of date2num it could be checked whether the first element of the array hasattr 'value'. This is true for pandas Timedelta/Timestamp/NaT. So there could be a separate conversion for these. 
The value of pandas.NaT is actually defined as the minimum value of numpy's int64. This can be used to identify and convert it. I have tried it and it is arguably much cleaner and certainly more concise than the current converter function I proposed.

Of course, if pandas nat simply stays unsupported, this is not necessary. Although I really think it would be nice if it were.

Reply:
I am fine with us trying to support NaT.  I just don't understand why its not the same as numpy's NaT.  It seems to me they just need two NaT's a datetime one and a timedelta one, and then we could make it work with no issues I think...

Reply:
I'm pretty sure that two NaT's won't happen for pandas. I skimmed over the roadmap and a linked proposal for a new type system in pandas. It looks like the place is to actually only have one "not-a-value" type for everything.

Reply:
> The value of pandas.NaT is actually defined as the minimum value of numpy's int64. This can be used to identify and convert it. I have tried it and it is arguably much cleaner and certainly more concise than the current converter function I proposed.

What does numpy do?  If its the same, then that is totally fine.  If its different, I guess there is a tiny risk someone will someday have a very large negative timedelta64, but I agree that risk seems small.

Reply:
As far as I can tell, numpy can/does not convert panda's nat.

Reply:
Sure but numpy has a timedelta64('NAT')

Reply:
At least internally they're doing the same

```
/* The special not-a-time (NaT) value */
#define NPY_DATETIME_NAT NPY_MIN_INT64
```